# Improving Employee Retention by Predicting Employee Attrition

Table of Content :

- Introducation
- Load Dataset
- Exploratory Data Analysis

## Introduction
**Background**

Sumber daya manusia (SDM) adalah aset utama yang perlu dikelola dengan baik oleh perusahaan agar tujuan bisnis dapat tercapai dengan efektif dan efisien. Pada kesempatan kali ini, kita akan menghadapi sebuah permasalahan tentang sumber daya manusia yang ada di perusahaan. Fokus kita adalah untuk mengetahui bagaimana cara menjaga karyawan agar tetap bertahan di perusahaan yang ada saat ini yang dapat mengakibatkan bengkaknya biaya untuk rekrutmen karyawan serta pelatihan untuk mereka yang baru masuk. Dengan mengetahui faktor utama yang menyebabkan karyawan tidak merasa, perusahaan dapat segera menanggulanginya dengan membuat program-program yang relevan dengan permasalahan karyawan.

**Goal:**

Meningkatkan retensi karyawan di perusahaan dengan mengidentifikasi dan mengatasi faktor-faktor utama yang menyebabkan ketidakpuasan serta meningkatkan keberlanjutan sumber daya manusia (SDM) sebagai aset utama perusahaan.

**Objectives:**


1. Mengidentifikasi faktor-faktor utama yang menyebabkan ketidakpuasan dan keengganan karyawan untuk tetap berada di perusahaan.
2. Membuat survei atau mekanisme umpan balik untuk mendapatkan pandangan langsung dari karyawan terkait dengan aspek-aspek yang dapat ditingkatkan.
3. Menganalisis hasil survei untuk menentukan tren dan pola dalam faktor-faktor ketidakpuasan yang ditemukan.
4. Mengembangkan program-program yang sesuai dan relevan untuk mengatasi setiap faktor ketidakpuasan yang diidentifikasi.
5. Menyusun strategi komprehensif untuk meningkatkan kepuasan karyawan dan meningkatkan retensi mereka.
6. Memonitor efektivitas program-program yang diimplementasikan melalui evaluasi rutin dan umpan balik dari karyawan.
7. Mengukur penurunan biaya rekrutmen dan pelatihan dengan memantau peningkatan retensi karyawan dan kepuasan mereka.
8. Meningkatkan komunikasi antara manajemen dan karyawan untuk memastikan pemahaman yang lebih baik tentang kebutuhan dan harapan karyawan.
9. Membentuk tim internal atau komite karyawan untuk berkolaborasi dalam merancang dan melaksanakan inisiatif retensi.
10. Menilai dan menyempurnakan program-program secara berkala berdasarkan umpan balik dan evaluasi kinerja perusahaan.

## Load Dataset

In [1]:
import pandas as pd 
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt 
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

In [2]:
df = pd.read_csv('Improving Employee Retention by Predicting Employee Attrition Using Machine Learning.csv')
df.sample(3)

Username  EnterpriseID StatusPernikahan JenisKelamin  \
39    blissfulHyena1        111140         Bercerai       Wanita   
207  troubledQuiche5        106480          Menikah       Wanita   
20    peacefulCoati5        100301          Menikah       Wanita   

    StatusKepegawaian                      Pekerjaan           JenjangKarir  \
39           FullTime        Digital Product Manager  Freshgraduate_program   
207          FullTime  Software Engineer (Front End)              Mid_level   
20           FullTime                Product Manager           Senior_level   

    PerformancePegawai     AsalDaerah HiringPlatform  ...         NomorHP  \
39               Bagus  Jakarta Timur         Indeed  ...  +6289713466xxx   
207       Sangat_bagus  Jakarta Pusat         Indeed  ...  +6285896075xxx   
20        Sangat_bagus  Jakarta Utara       LinkedIn  ...  +6283355427xxx   

                             Email  TingkatPendidikan  PernahBekerja  \
39   blissfulHyena1934@hotmail.com            Sarjana              1   
207  troubledQuiche5924@icloud.com            Sarjana              1   
20     peacefulCoati5154@yahoo.com            Sarjana              1   

     IkutProgramLOP   AlasanResign TanggalLahir TanggalHiring  \
39              NaN  masih_bekerja   1982-03-28     2013-8-19   
207             NaN     leadership   1972-11-21    2017-01-07   
20              NaN  masih_bekerja   1977-05-22    2012-01-09   

    TanggalPenilaianKaryawan  TanggalResign  
39                2020-02-12              -  
207               2020-02-06      2013-5-30  
20                2020-01-07              -  

[3 rows x 25 columns]

## Exploratory Data Analysis

### Overview Data

Melakukan investigasi awal terhadap data untuk memahami karakteristik dan informasi yang terkandung didalamnya termasuk missing value dan duplikat.

In [3]:
df.shape

(287, 25)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 287 entries, 0 to 286
Data columns (total 25 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Username                            287 non-null    object 
 1   EnterpriseID                        287 non-null    int64  
 2   StatusPernikahan                    287 non-null    object 
 3   JenisKelamin                        287 non-null    object 
 4   StatusKepegawaian                   287 non-null    object 
 5   Pekerjaan                           287 non-null    object 
 6   JenjangKarir                        287 non-null    object 
 7   PerformancePegawai                  287 non-null    object 
 8   AsalDaerah                          287 non-null    object 
 9   HiringPlatform                      287 non-null    object 
 10  SkorSurveyEngagement                287 non-null    int64  
 11  SkorKepuasanPegawai                 282 non-n

In [5]:
df.isnull().sum()

Username                                0
EnterpriseID                            0
StatusPernikahan                        0
JenisKelamin                            0
StatusKepegawaian                       0
Pekerjaan                               0
JenjangKarir                            0
PerformancePegawai                      0
AsalDaerah                              0
HiringPlatform                          0
SkorSurveyEngagement                    0
SkorKepuasanPegawai                     5
JumlahKeikutsertaanProjek               3
JumlahKeterlambatanSebulanTerakhir      1
JumlahKetidakhadiran                    6
NomorHP                                 0
Email                                   0
TingkatPendidikan                       0
PernahBekerja                           0
IkutProgramLOP                        258
AlasanResign                           66
TanggalLahir                            0
TanggalHiring                           0
TanggalPenilaianKaryawan          

Terdapat beberapa fitur yang missing value yaitu `SkorKepuasanPegawai`, `JumlahKeikutsertaanProjek`, `JumlahKeterlambatanSebulanTerakhir`, `JumlahKetidakhadiran`, `IkutProgramLOP`, `AlasanResign`

In [6]:
# overview
list_item = []
for col in df.columns:
    list_item.append([col, df[col].dtype, df[col].isna().sum(), round(100*df[col].isna().sum()/len(df[col]), 2), df[col].nunique(), df[col].unique()[:10]])
desc_df = pd.DataFrame(data=list_item, columns='Username, EnterpriseID, StatusPernikahan, JenisKelamin, StatusKepegawaian, Pekerjaan'.split(","))
desc_df

Username  EnterpriseID   StatusPernikahan  \
0                             Username        object                  0   
1                         EnterpriseID         int64                  0   
2                     StatusPernikahan        object                  0   
3                         JenisKelamin        object                  0   
4                    StatusKepegawaian        object                  0   
5                            Pekerjaan        object                  0   
6                         JenjangKarir        object                  0   
7                   PerformancePegawai        object                  0   
8                           AsalDaerah        object                  0   
9                       HiringPlatform        object                  0   
10                SkorSurveyEngagement         int64                  0   
11                 SkorKepuasanPegawai       float64                  5   
12           JumlahKeikutsertaanProjek       float64                  3   
13  JumlahKeterlambatanSebulanTerakhir       float64                  1   
14                JumlahKetidakhadiran       float64                  6   
15                             NomorHP        object                  0   
16                               Email        object                  0   
17                   TingkatPendidikan        object                  0   
18                       PernahBekerja        object                  0   
19                      IkutProgramLOP       float64                258   
20                        AlasanResign        object                 66   
21                        TanggalLahir        object                  0   
22                       TanggalHiring        object                  0   
23            TanggalPenilaianKaryawan        object                  0   
24                       TanggalResign        object                  0   

     JenisKelamin   StatusKepegawaian  \
0            0.00                 285   
1            0.00                 287   
2            0.00                   5   
3            0.00                   2   
4            0.00                   3   
5            0.00                  14   
6            0.00                   3   
7            0.00                   5   
8            0.00                   5   
9            0.00                   9   
10           0.00                   5   
11           1.74                   5   
12           1.05                   9   
13           0.35                   7   
14           2.09                  22   
15           0.00                 287   
16           0.00                 287   
17           0.00                   3   
18           0.00                   2   
19          89.90                   2   
20          23.00                  11   
21           0.00                 284   
22           0.00                  97   
23           0.00                 127   
24           0.00                  53   

                                            Pekerjaan  
0   [spiritedPorpoise3, jealousGelding2, pluckyMue...  
1   [111065, 106080, 106452, 106325, 111171, 10641...  
2      [Belum_menikah, Menikah, Bercerai, Lainnya, -]  
3                                      [Pria, Wanita]  
4                   [Outsource, FullTime, Internship]  
5   [Software Engineer (Back End), Data Analyst, S...  
6    [Freshgraduate_program, Senior_level, Mid_level]  
7   [Sangat_bagus, Sangat_kurang, Bagus, Biasa, Ku...  
8   [Jakarta Timur, Jakarta Utara, Jakarta Pusat, ...  
9   [Employee_Referral, Website, Indeed, LinkedIn,...  
10                                    [4, 3, 2, 1, 5]  
11                     [4.0, 3.0, 5.0, nan, 2.0, 1.0]  
12  [0.0, 4.0, 6.0, nan, 7.0, 3.0, 5.0, 1.0, 2.0, ...  
13           [0.0, 4.0, 3.0, 5.0, 2.0, 6.0, 1.0, nan]  
14  [9.0, 3.0, 11.0, 6.0, 10.0, 19.0, 4.0, 5.0, 2....  
15  [+6282232522xxx, +6281270745xxx, +6281346215xx...  
16  [spiritedPorpoise3135@yahoo.com, jealousGeldin...  
17  

Di Fitur pernah bekerja terdapat keanehan data yaitu data numerik dengan kategorikal contohnya [1, yes]

In [7]:
df.duplicated().sum()

0

In [8]:
# numerikal fitur
nums = df.loc[:, (df.dtypes == 'int64') | (df.dtypes == float)].columns.tolist()
nums

['EnterpriseID',
 'SkorSurveyEngagement',
 'SkorKepuasanPegawai',
 'JumlahKeikutsertaanProjek',
 'JumlahKeterlambatanSebulanTerakhir',
 'JumlahKetidakhadiran',
 'IkutProgramLOP']

In [9]:
# kategorikal fitur
cats = df.loc[:, (df.dtypes == 'object')].columns.tolist()
cats

['Username',
 'StatusPernikahan',
 'JenisKelamin',
 'StatusKepegawaian',
 'Pekerjaan',
 'JenjangKarir',
 'PerformancePegawai',
 'AsalDaerah',
 'HiringPlatform',
 'NomorHP',
 'Email',
 'TingkatPendidikan',
 'PernahBekerja',
 'AlasanResign',
 'TanggalLahir',
 'TanggalHiring',
 'TanggalPenilaianKaryawan',
 'TanggalResign']

## **Poin-poin Penting**

**Karakteristik Data:**

- Dataset terdiri dari 287 baris dan 25 fitur.
- 7 fitur numerik dan 18 fitur kategori

**Missing Value:**

- Terdapat missing value pada fitur :
    - SkorKepuasanPegawai 
    - JumlahKeikutsertaanProjek
    - JumlahKeterlambatanSebulanTerakhir
    - JumlahKetidakhadiran
    - IkutProgramLOP
    - AlasanResign

 **Duplikat Data:**
- Tidak ada duplikat data dalam dataset.

## Statistical Summary

In [10]:
df[nums].describe().round(0).T

count      mean     std       min  \
EnterpriseID                        287.0  105923.0  4045.0  100282.0   
SkorSurveyEngagement                287.0       3.0     1.0       1.0   
SkorKepuasanPegawai                 282.0       4.0     1.0       1.0   
JumlahKeikutsertaanProjek           284.0       1.0     2.0       0.0   
JumlahKeterlambatanSebulanTerakhir  286.0       0.0     1.0       0.0   
JumlahKetidakhadiran                281.0      10.0     7.0       1.0   
IkutProgramLOP                       29.0       1.0     1.0       0.0   

                                         25%       50%       75%       max  
EnterpriseID                        101269.0  106069.0  110514.0  111703.0  
SkorSurveyEngagement                     3.0       3.0       4.0       5.0  
SkorKepuasanPegawai                      3.0       4.0       5.0       5.0  
JumlahKeikutsertaanProjek                0.0       0.0       0.0       8.0  
JumlahKeterlambatanSebulanTerakhir       0.0       0.0       0.0       6.0  
JumlahKetidakhadiran                     5.0      10.0      15.0      55.0  
IkutProgramLOP                           0.0       1.0       1.0       1.0

**Key Takeaways:**

- Dilihat dari nilai mean dan median yang tidak begitu jauh nilainya, distribusi data cenderung mendekati normal
- Mayoritas karyawan memiliki **SkorSurveyEngagement 3.0 dari 5.0**
- Mayoritas karyawan memiliki **SkorKepuasanPegawai 4.0 dari 5.0**
- Mayoritas karyawan memiliki **JumlahKeikutsertaanProjek 0.0 dari 8.0**
- Mayoritas karyawan memiliki **JumlahKeterlambatanSebulanTerakhir	0.0 dari 6.0**
- Mayoritas karyawan memiliki **JumlahKetidakhadiran 10.0 dari 55.0**
- Mayoritas karyawan memiliki **IkutProgramLOP 1.0 dari 1.0 dari jumlah karyawan yang totalnya 287 yang hanya mengisi 29**

**Penjelasan**
- Jika mayoritas karyawan memiliki **Skor Survey Engagement sebesar 3.0 dari 5.0**, hal ini menunjukkan adanya tingkat keterlibatan yang cenderung rata-rata. 
- Jika **Skor Kepuasan Pegawai sebesar 4.0 dari 5.0**, itu menunjukkan bahwa mayoritas pegawai merasa cukup puas dengan kondisi dan pengalaman mereka di perusahaan. Meskipun sudah mencapai tingkat kepuasan yang baik, perusahaan dapat tetap mengambil langkah-langkah untuk mempertahankan dan meningkatkan kepuasan pegawai
- Jika mayoritas karyawan memiliki Jumlah **Keikutsertaan Projek sebesar 0.0 dari 8.0**, ini mungkin menunjukkan bahwa karyawan tidak terlibat dalam proyek-proyek perusahaan atau memiliki tingkat partisipasi yang sangat rendah. Situasi ini dapat memiliki dampak negatif pada motivasi, pengembangan keterampilan, dan rasa kepemilikan terhadap pekerjaan.
- Jika mayoritas karyawan memiliki **Jumlah Keterlambatan sebesar 0.0 dari 6.0**, itu menunjukkan bahwa sebagian besar karyawan memiliki rekam jejak kehadiran yang baik dan jarang mengalami keterlambatan. Ini bisa diartikan sebagai indikator positif terhadap disiplin dan keteraturan karyawan dalam memenuhi kewajiban mereka terkait waktu kerja. Walaupun demikian, tetap diperlukan pemantauan dan tindakan yang proaktif untuk memastikan bahwa keadaan ini tetap terjaga.
- Jika mayoritas karyawan memiliki **Jumlah Ketidakhadiran sebesar 10.0 dari 55.0**, ini menunjukkan bahwa ada sejumlah tingkat ketidakhadiran di antara karyawan yang mungkin memerlukan perhatian lebih lanjut. Ketidakhadiran dapat memiliki dampak negatif pada produktivitas dan kinerja perusahaan secara keseluruhan. 
- Jika mayoritas dari karyawan, yang jumlahnya total 287 namun hanya 29 yang mengisi, memiliki Skor Ikut Program Loyalitas dan Organisasi (LOP) sebesar 1.0 dari 1.0, hal ini menunjukkan bahwa karyawan yang telah mengisi survei sangat positif terhadap dan berpartisipasi dalam program LOP yang ditawarkan oleh perusahaan. Meskipun jumlah responden relatif kecil dibandingkan dengan jumlah total karyawan, skor penuh menunjukkan tingkat kepuasan yang tinggi terhadap program tersebut. 

In [11]:
df[cats].describe().T

count unique                             top freq
Username                   287    285                      boredEggs0    2
StatusPernikahan           287      5                   Belum_menikah  132
JenisKelamin               287      2                          Wanita  167
StatusKepegawaian          287      3                        FullTime  217
Pekerjaan                  287     14    Software Engineer (Back End)  109
JenjangKarir               287      3           Freshgraduate_program  169
PerformancePegawai         287      5                           Biasa   85
AsalDaerah                 287      5                   Jakarta Pusat   72
HiringPlatform             287      9                          Indeed   85
NomorHP                    287    287                  +6282232522xxx    1
Email                      287    287  spiritedPorpoise3135@yahoo.com    1
TingkatPendidikan          287      3                         Sarjana  176
PernahBekerja              287      2                               1  286
AlasanResign               221     11                   masih_bekerja  132
TanggalLahir               287    284                      1987-06-14    2
TanggalHiring              287     97                      2011-01-10   12
TanggalPenilaianKaryawan   287    127                       2020-1-14   17
TanggalResign              287     53                               -  198

Berikut adalah simpulan singkat dari data yang disediakan:

1. **Username:**
    - Terdapat 287 entri, dengan 285 nama pengguna unik.
    - Nama pengguna 'boredEggs0' muncul 2 kali.

2. **Status Pernikahan:**
    - Ada 5 status pernikahan yang berbeda.
    - Mayoritas karyawan (132) belum menikah.

3. **Jenis Kelamin:**
    - Terdapat dua jenis kelamin, dengan mayoritas karyawan (167) adalah wanita.

4. **Status Kepegawaian:**
    - Ada 3 status kepegawaian, dengan mayoritas karyawan (217) bekerja sebagai karyawan penuh waktu (FullTime).

5. **Pekerjaan:**
    - Terdapat 14 jenis pekerjaan berbeda.
    - Posisi Software Engineer (Back End) paling banyak diisi, dengan 109 karyawan.

6. **Jenjang Karir:**
    - Ada 3 jenjang karir yang berbeda.
    - Mayoritas karyawan (169) berada pada jenjang Freshgraduate_program.

7. Performance Pegawai:
    - Terdapat 5 level performa pegawai.
    - Mayoritas karyawan (85) memiliki performa biasa.

8. **Asal Daerah:**
    - Ada 5 asal daerah yang berbeda.
    - Jakarta Pusat merupakan asal daerah mayoritas dengan 72 karyawan.

9. **Hiring Platform:**
    - Terdapat 9 platform perekrutan yang berbeda.
    - Indeed adalah platform perekrutan yang paling sering digunakan oleh karyawan (85 kali).

10. **Nomor HP dan Email:**
    - Setiap karyawan memiliki nomor HP dan alamat email yang unik.

11. **Tingkat Pendidikan:**
    - Ada 3 tingkat pendidikan yang berbeda.
    - Mayoritas karyawan (176) memiliki gelar Sarjana.

12. **Pernah Bekerja:**
    - Mayoritas karyawan (286 dari 287) telah bekerja sebelumnya.

13. **Alasan Resign:**
    - Terdapat 11 alasan resign yang berbeda.
    - Alasan resign paling umum adalah masih bekerja (132 karyawan).

14. **Tanggal Lahir:**
    - Terdapat 287 tanggal lahir yang berbeda, dengan beberapa tanggal lahir yang sama muncul 2 kali.

15. **Tanggal Hiring:**
    - Terdapat 97 tanggal hiring yang berbeda, dengan beberapa tanggal hiring yang muncul 12 kali.

16. **Tanggal Penilaian Karyawan:**
    - Terdapat 127 tanggal penilaian karyawan yang berbeda, dengan beberapa tanggal penilaian yang muncul 17 kali.

17. **Tanggal Resign:**
    - Terdapat 53 tanggal resign yang berbeda, dengan mayoritas (198 kali) tidak memiliki tanggal resign yang diisi (-).

## Data preprocessing
Sebelum memulai pemrosesan data, tahap paling awal yang harus dilakukan adalah mempersiapkan data mentah menjadi data yang bersih dan siap diolah. Kamu harus melakukan data preprocessing, yakni menangani berbagai permasalahan data seperti data yang kosong, data yang tidak sesuai, hingga mengidentifikasi data-data yang tidak dibutuhkan.

In [12]:
#Mengatasi data null 
df['SkorKepuasanPegawai'].fillna(df['SkorKepuasanPegawai'].median(), inplace=True)
df['JumlahKeikutsertaanProjek'].fillna(df['JumlahKeikutsertaanProjek'].mean(), inplace=True)
df['JumlahKeterlambatanSebulanTerakhir'].fillna(df['JumlahKeterlambatanSebulanTerakhir'].mean(), inplace=True)
df['JumlahKetidakhadiran'].fillna(df['JumlahKetidakhadiran'].median(), inplace=True)
df['IkutProgramLOP'].fillna(df['IkutProgramLOP'].mode()[0], inplace=True)
df['AlasanResign'].fillna('Tidak Diketahui', inplace=True)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 287 entries, 0 to 286
Data columns (total 25 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Username                            287 non-null    object 
 1   EnterpriseID                        287 non-null    int64  
 2   StatusPernikahan                    287 non-null    object 
 3   JenisKelamin                        287 non-null    object 
 4   StatusKepegawaian                   287 non-null    object 
 5   Pekerjaan                           287 non-null    object 
 6   JenjangKarir                        287 non-null    object 
 7   PerformancePegawai                  287 non-null    object 
 8   AsalDaerah                          287 non-null    object 
 9   HiringPlatform                      287 non-null    object 
 10  SkorSurveyEngagement                287 non-null    int64  
 11  SkorKepuasanPegawai                 287 non-n

In [14]:
# overview
list_item = []
for col in df.columns:
    list_item.append([col, df[col].dtype, df[col].isna().sum(), round(100*df[col].isna().sum()/len(df[col]), 2), df[col].nunique(), df[col].unique()[:10]])
desc_df = pd.DataFrame(data=list_item, columns='Username, EnterpriseID, StatusPernikahan, JenisKelamin, StatusKepegawaian, Pekerjaan'.split(","))
desc_df

Username  EnterpriseID   StatusPernikahan  \
0                             Username        object                  0   
1                         EnterpriseID         int64                  0   
2                     StatusPernikahan        object                  0   
3                         JenisKelamin        object                  0   
4                    StatusKepegawaian        object                  0   
5                            Pekerjaan        object                  0   
6                         JenjangKarir        object                  0   
7                   PerformancePegawai        object                  0   
8                           AsalDaerah        object                  0   
9                       HiringPlatform        object                  0   
10                SkorSurveyEngagement         int64                  0   
11                 SkorKepuasanPegawai       float64                  0   
12           JumlahKeikutsertaanProjek       float64                  0   
13  JumlahKeterlambatanSebulanTerakhir       float64                  0   
14                JumlahKetidakhadiran       float64                  0   
15                             NomorHP        object                  0   
16                               Email        object                  0   
17                   TingkatPendidikan        object                  0   
18                       PernahBekerja        object                  0   
19                      IkutProgramLOP       float64                  0   
20                        AlasanResign        object                  0   
21                        TanggalLahir        object                  0   
22                       TanggalHiring        object                  0   
23            TanggalPenilaianKaryawan        object                  0   
24                       TanggalResign        object                  0   

     JenisKelamin   StatusKepegawaian  \
0             0.0                 285   
1             0.0                 287   
2             0.0                   5   
3             0.0                   2   
4             0.0                   3   
5             0.0                  14   
6             0.0                   3   
7             0.0                   5   
8             0.0                   5   
9             0.0                   9   
10            0.0                   5   
11            0.0                   5   
12            0.0                  10   
13            0.0                   8   
14            0.0                  22   
15            0.0                 287   
16            0.0                 287   
17            0.0                   3   
18            0.0                   2   
19            0.0                   2   
20            0.0                  12   
21            0.0                 284   
22            0.0                  97   
23            0.0                 127   
24            0.0                  53   

                                            Pekerjaan  
0   [spiritedPorpoise3, jealousGelding2, pluckyMue...  
1   [111065, 106080, 106452, 106325, 111171, 10641...  
2      [Belum_menikah, Menikah, Bercerai, Lainnya, -]  
3                                      [Pria, Wanita]  
4                   [Outsource, FullTime, Internship]  
5   [Software Engineer (Back End), Data Analyst, S...  
6    [Freshgraduate_program, Senior_level, Mid_level]  
7   [Sangat_bagus, Sangat_kurang, Bagus, Biasa, Ku...  
8   [Jakarta Timur, Jakarta Utara, Jakarta Pusat, ...  
9   [Employee_Referral, Website, Indeed, LinkedIn,...  
10                                    [4, 3, 2, 1, 5]  
11                          [4.0, 3.0, 5.0, 2.0, 1.0]  
12  [0.0, 4.0, 6.0, 1.1795774647887325, 7.0, 3.0, ...  
13  [0.0, 4.0, 3.0, 5.0, 2.0, 6.0, 1.0, 0.41258741...  
14  [9.0, 3.0, 11.0, 6.0, 10.0, 19.0, 4.0, 5.0, 2....  
15  [+6282232522xxx, +6281270745xxx, +6281346215xx...  
16  [spiritedPorpoise3135@yahoo.com, jealousGeldin...  
17  

In [15]:
# Menghapus kolom "nomorhp" karena memiliki satu unique value
df.drop("NomorHP", axis=1, inplace=True)

# Menghapus kolom "email" karena memiliki satu unique value
df.drop("Email", axis=1, inplace=True)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 287 entries, 0 to 286
Data columns (total 23 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Username                            287 non-null    object 
 1   EnterpriseID                        287 non-null    int64  
 2   StatusPernikahan                    287 non-null    object 
 3   JenisKelamin                        287 non-null    object 
 4   StatusKepegawaian                   287 non-null    object 
 5   Pekerjaan                           287 non-null    object 
 6   JenjangKarir                        287 non-null    object 
 7   PerformancePegawai                  287 non-null    object 
 8   AsalDaerah                          287 non-null    object 
 9   HiringPlatform                      287 non-null    object 
 10  SkorSurveyEngagement                287 non-null    int64  
 11  SkorKepuasanPegawai                 287 non-n

## Annual Report On Employee Number Changes
Dalam konteks manajemen karyawan, perusahaan perlu mengetahui berapa perubahan jumlah karyawan setiap tahunnya agar bisa menyesuaikan langkah strategis guna meningkatkan kinerja perusahaan dan bisa bertahan di dunia yang terdisrupsi ini. Salah satu langkah yang telah dibuat perusahaan adalah melakukan perubahan budaya kerja setiap tahunnya yang mengacu pada jumlah karyawan yang ada, semakin banyak karyawan maka office space yang perlu disiapkan juga semakin bertambah, serta kebijakan-kebijakan lain seperti besaran bonus maksimal untuk karyawan yang performanya baik di tahun tersebut dan lain-lain. Setiap tahunnya perusahaan menyadari ada yang keluar dan ada yang masuk (karyawan), namun apabila pada tahun tertentu jumlah karyawan yang keluar lebih banyak dari yang masuk artinya perusahaan perlu berbenah diri agar karyawan lain tetap bertahan sehingga rancangan strategi jangka panjang tetap dapat dilaksanakan tanpa kendala.

In [17]:
df['StatusPernikahan'] = df['StatusPernikahan'].replace('-','')

df['TanggalLahir'] =  pd.to_datetime(df['TanggalLahir'])

df['TahunLahir'] = df['TanggalLahir'].dt.year

df['Umur'] = 2024 - df['TahunLahir']

df['TanggalResign'] = df['TanggalResign'].replace('-', np.nan)

df['TanggalResign'] = pd.to_datetime(df['TanggalResign'])
df['TanggalHiring'] = pd.to_datetime(df['TanggalHiring'])

def check_resign(date):
    if pd.isnull(date):
        return 'No'
    else:
        return 'Yes'
    
df['Resign'] = df['TanggalResign'].apply(check_resign)

df['TahunHiring'] = df['TanggalHiring'].dt.year
df['TahunResign'] = df['TanggalResign'].dt.year

In [18]:
annual_hiring_agg =  (
    df
    .groupby('TahunHiring')['Username']
    .count()
    .to_frame()
    .reset_index()
    .rename(columns={'Username': 'JumlahPegawaiHire'})
    .sort_values(by='TahunHiring', ascending=True)
)

annual_resign_agg = (
    df
    .groupby('TahunResign')['Username']
    .count()
    .to_frame()
    .reset_index()
    .rename(columns={'Username': 'JumlahPegawaiResign'})
    .sort_values(by='TahunResign', ascending=True)
)

annual_hr_agg = annual_hiring_agg.merge(annual_resign_agg, how='outer', left_on='TahunHiring', right_on='TahunResign')
annual_hr_agg['TahunHiring'] = annual_hr_agg['TahunHiring'].fillna(annual_hr_agg['TahunResign'])
annual_hr_agg['TahunResign'] = annual_hr_agg['TahunResign'].fillna(annual_hr_agg['TahunHiring'])
annual_hr_agg.fillna(0, inplace=True)

annual_hr_agg['PegawaiBertahan'] = annual_hr_agg['JumlahPegawaiHire'] - annual_hr_agg['JumlahPegawaiResign']
annual_hr_agg

TahunHiring  JumlahPegawaiHire  TahunResign  JumlahPegawaiResign  \
0        2006.0                1.0       2006.0                  0.0   
1        2007.0                2.0       2007.0                  0.0   
2        2008.0                2.0       2008.0                  0.0   
3        2009.0                7.0       2009.0                  0.0   
4        2010.0                8.0       2010.0                  0.0   
5        2011.0               76.0       2011.0                  0.0   
6        2012.0               41.0       2012.0                  0.0   
7        2013.0               43.0       2013.0                  5.0   
8        2014.0               56.0       2014.0                 12.0   
9        2015.0               31.0       2015.0                  8.0   
10       2016.0               14.0       2016.0                  8.0   
11       2017.0                5.0       2017.0                 19.0   
12       2018.0                1.0       2018.0                 26.0   
13       2019.0                0.0       2019.0                  5.0   
14       2020.0                0.0       2020.0                  6.0   

    PegawaiBertahan  
0               1.0  
1               2.0  
2               2.0  
3               7.0  
4               8.0  
5              76.0  
6              41.0  
7              38.0  
8              44.0  
9              23.0  
10              6.0  
11            -14.0  
12            -25.0  
13             -5.0  
14             -6.0

In [24]:
import plotly.graph_objects as go
dfg = annual_hr_agg

dfg['Net'] = dfg['JumlahPegawaiHire'] - dfg['JumlahPegawaiResign']

measures = ['relative' if net >= 0 else 'relative' for net in dfg['Net']]

fig = go.Figure(go.Waterfall(
    name="20", orientation="v",
    measure=measures,  # Corrected property name
    x=dfg['TahunHiring'].to_list(),
    textposition="outside",
    text=dfg['Net'].to_list(),
    y=dfg['Net'].to_list(),
    decreasing={"marker": {"color": "Red"}},
    increasing={"marker": {"color": "Green"}},
))

fig.update_layout(title="Perubahan Jumlah Pegawai", showlegend=False)

fig.show()

In [25]:
import plotly.graph_objects as go
from plotly.offline import iplot

dfg = annual_hr_agg

dfg['Net'] = dfg['JumlahPegawaiHire'] - dfg['JumlahPegawaiResign']

measures = ['relative' if net >= 0 else 'relative' for net in dfg['Net']]

fig = go.Figure(go.Waterfall(
    name="20", orientation="v",
    measure=measures,  # Corrected property name
    x=dfg['TahunHiring'].to_list(),
    textposition="outside",
    text=dfg['Net'].to_list(),
    y=dfg['Net'].to_list(),
    decreasing={"marker": {"color": "Red"}},
    increasing={"marker": {"color": "Green"}},
))

fig.update_layout(title="Perubahan Jumlah Pegawai", showlegend=False)

# Display the figure in the Jupyter Notebook
iplot(fig)